### ISSUE TITLE
[discuss] Let's talk about refs

### ISSUE THREAD

**Issue by lookfirst** [mikro-orm/mikro-orm#145]
Refs are great because they make it possible for me to not load objects when I don't need to. They are central to most good ORM's because they provide a way to optimize performance without a lot of contextual overload. Right now the api is a bit weird and I think refs should become more of a first class citizen in mikro. For example:

`(await thing1!.createdBy.load()).email`
(assuming findOne(..., ..., {populate...}) ... `thing1!.createdBy.unwrap().email`

This is um, ugly. Now I'm encoding the reference api into how I access things which seems counter intuitive to a good ORM and makes migrating between non-refs to refs (and the other way around) a bit more tricky.

The first example isn't entirely bad. I prefer `.get()` over `.load()` though. But maybe that is from my experience with [Objectify](https://github.com/objectify/objectify/wiki/Entities#refs).

For the second example, what if you had generated a `get` accessor for the email property? It could look up the tree to unwrap things automatically. Not sure how to accomplish this in JS, but Proxy objects or something? Could also consider the [Ofy @Load](https://github.com/objectify/objectify/wiki/Entities#load) way of doing things.

There is another bit of experience, which is [hiding refs](https://github.com/objectify/objectify/wiki/Entities#hiding-ref).

**Comment by lookfirst** [comment]
I also think that `.init()` is weird too. 

I feel like all this can be consolidated into a more cohesive api.

**Comment by B4nan** [comment]
> This is um, ugly. Now I'm encoding the reference api into how I access things which seems counter intuitive to a good ORM and makes migrating between non-refs to refs (and the other way around) a bit more tricky.

Well the purpose of `Reference` wrapper is to force you do this so your code is type safe and guaranteed to work. You don't have to use it obviously... 

> The first example isn't entirely bad. I prefer .get() over .load() though. But maybe that is from my experience with Objectify.

I don't agree, `get` sounds more like `unwrap` than `load`. `unwrap/load` are to me more expressive. 

> For the second example, what if you had generated a get accessor for the email property?

That was my original idea, but it does not work like this in JS. User defines properties in entity, so I can't override them with native get/set pair. At least I don't know how currently :/

> I also think that .init() is weird too.

Well, it feels weird now that we have those things next to each other. Originally there was no `Reference` wrapper, that is quite new thing. A reference was previously called just a simple entity that has only PK defined. 

Btw thanks for opening this, I was thinking about the same too recently.

**Comment by B4nan** [comment]
What I was thinking about recently is to add `load` and `unwrap` methods to all entities so their API would be compatible with `Reference` wrapper. This would allow users to do:

- assigning entity (no matter if initialized or not) to field defined as reference
- returning populated entities from find/findOne methods without wrapping them in the wrapper

Another idea how to improve what we have now - although I cannot define getter for all properties on entity, I could define them on the `Reference` wrapper. But still, getters are synchronous, so I cannot lazy load things there (that is why we have that `unwrap/load` methods). That would allow user to work with the reference as with normal entity, if it was already populated.

**Comment by lookfirst** [comment]
> although I cannot define getter for all properties on entity

Have you seen JS Proxy object API? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/handler/get

**Comment by B4nan** [comment]
I know about it but never used it... :)

**Comment by lookfirst** [comment]
I think it is the magic that you are looking for.

**Comment by B4nan** [comment]
> I think it is the magic that you are looking for.

Right, now it makes sense! :]

tagging here also @vimmerru as he was requesting this `Reference` wrapper, also linking the issue original #107.

**Comment by B4nan** [comment]
Still what I said applies, if we add those getters to the wrapper, it will have same interface as the entity itself so it does not serves the purpose why it was added (to improve type safety). It could work this way only if we are able to run async code in the getter, which is obviously not possible as getters are synchronous. 

Although they can return promise which you could awaited...

**Comment by lookfirst** [comment]
I like promises.

**Comment by vimmerru** [comment]
In my vision good ts ORM should provide at least:

- Static checking of model correctness
- Static checking of queries correctness

The current approach with ```isInitialized``` unfortunately breaks static-level correctness checking. Ref type help here a bit, but keep a lot of problems related to queries and model static checking.

@B4nan I have an early idea how to achieve good static checking for population based on mapped types feature of ts and keep static checking of queries.

**Comment by vimmerru** [comment]
Here is an example of function that resolves all promises in object by keeping typing:

```
type UnPromisifiedObject<T> = {[k in keyof T]: UnPromisify<T[k]>}
type UnPromisify<T> = T extends Promise<infer U> ? U : T;

async function promise_props<T extends {[key: string]: Promise<any>}>(obj: T): Promise<UnPromisifiedObject<T>> {
    const keys = Object.keys(obj);
    const awaitables = keys.map(key => obj[key]);

    const values = await Promise.all(awaitables);
    const result = {} as any;

    keys.forEach((key, i) => {
        result[key] = values[i];
    });
    return result as UnPromisifiedObject<T>;
}

async function main() {
    const x = {
        company: Promise.resolve("company"),
        page: Promise.resolve(1)
    };
    const res = await promise_props(x);
    const company = res.company;  // company is a string here
    const page = res.page;        // page is a number here
}
```

**Comment by stephenh** [comment]
Fwiw, I think this:

> That would allow user to work with the reference as with normal entity, if it was already populated.

```typescript
const thing = await em.findOneOrFail(Thing, 1, ['createdBy']); 
console.log(thing.createdBy.email); // no unwrap needed
```

Should be a non-goal, because it's coupling the behavior of `thing.createdBy.email` to "did you or did you not pass `['createdBy']` when loading `Thing 1`".

IMO a big point of ORMs is that you can implement business logic in one spot (i.e. lifecycle hooks) that is generally agnostic of "how did this data get loaded?" w/o also constantly refetching the same row from the db (i.e. because of the identity map).

However, when `Thing` is modeled as `@ManyToOne createdBy: User`, every single code path should really be doing an `await thing.createdBy.load()`, and the fact that Mikro doesn't enforce this seems like a mistake, as you're setting up users of Mikro to have a known class of errors (accessing properties that aren't initialized yet).

Basically, I see the `IdentifiedRef` style, i.e. `Thing` has a `@ManyToOne createdBy: IdentifiedRef<User>` and think that MikroORM should double-down on that as the "one true way" of modeling refs and actively deprecate the "looks spiffy b/c it works in two-line examples but will not scale to a reasonably-sized application" approach of `@ManyToOne createdBy: User`.

**Comment by B4nan** [comment]
Agreed with your points, the whole `Reference` wrapper is pretty new thing and I am not 100% satisfied with how it works, but in the long run it will be most probably the default way of handling references. I would say this (together with the #214) will be the main thing to resolve in v3 (or v4 as it might bring some breaking changes). 

I do not want to postpone the release forever, but if I will be able to at least find a way to throw errors when user tries to access properties of uninitialized entity, I'd love to have that part of v3. 

> (Also, I suppose pedantically/tangentially, non-Node/non-JS ORMs didn't have to struggle with this decision b/c the thing.createdBy call could be both synchronous-to-the-language but blocking-at-runtime, i.e. the ORM would realize "oh wait, this User is not loaded yet, and load it before returning from the createdBy getter. (Or, alternatively, it would return a ref from createdBy and then do the actual db load in the .email accessor, but again, it's sneaking "conditional I/O" into a "maybe blocking/maybe not" method call in Java/C#/etc.) Given that Node/JS just can't do this, any method that might possibly be async has to always be treated as async, I think users of Node/JS ORMs just have to accept that const email = (await thing.createdBy.get()).email is the only "it might be loaded/it might not", i.e. lazy-loading ORM, way of traversing the object graph, and just generally accept that fact of life, even if its slightly less pretty than the Java/C#/etc. alternatives. But, hey, at least its a promise and not a callback. :-) )

Yeah agreed on that as well. Btw you can also do `const email = await thing.createdBy.get('email')` to access the property directly to make it less ugly :]

**Comment by vimmerru** [comment]
@stephenh 

> IMO a big point of ORMs is that you can implement business logic in one spot (i.e. lifecycle hooks) that is generally agnostic of "how did this data get loaded?" w/o also constantly refetching the same row from the db (i.e. because of the identity map).

It contradicts my experience. You just can't avoid persistence logic in your app. For any non-trivial app you just can't avoid thinking on how data loaded maybe except some near single-user use cases.

Also i see tendency to use persistence model as some kind of 'universal' model. People use this for REST interfaces, json serialization, calling microservices and so on. This approach also doesn't scale.

The main idea of having IdentifiedReference is to show that entity reference and loaded entity are really different things.

Having a code that will automatically load entity parts when some business logic requires this works only for trivial cases.

**Comment by B4nan** [comment]
> Having a code that will automatically load entity parts when some business logic requires this works only for trivial cases.

Why do you think so? Have you seen #266? The proposal is to use dataloader "pattern" (or how to call it) that would allow to do batch loading of referenced entities. So the performance standpoint should be covered. Of course you sometimes need to load things conditionally, not just by ID (and you still can do it when needed), but in general you won't need to care about how relations are loaded. 

The proposal (as I understand it) would be pretty much automation of the `populate` parameter. 

Also for more complex cases, I am planning to add support for filters (sometimes called association scopes) which would allow you to apply additional conditions to the queries loading your relations (e.g. ignoring soft deleted entities). Another thing I want to focus during 3.x development.

**Comment by lookfirst** [comment]
I'll ping Objectify again here (it is mentioned at the very top)....

https://github.com/objectify/objectify/wiki/Entities#refs

**Comment by vimmerru** [comment]
> Why do you think so? Have you seen #266?

In general idea of DataLoader is good to fit tin GraphQL case, but i don't understand how it can help to load level-2 or level-3 entities or avoid loading of big referenced collections.

**Comment by darkbasic** [comment]
```ts
type DeRef<T> = T extends Ref<infer U> ? U : T;

type DeRefObject<T, > = {[k in keyof T]: DeRef<T[k]>}
```

@vimmerru I think this is great. `em.find` could implement some kind of `DeRefObjectPartial` so that it automatically unwraps the fields you ask him to populate:

```ts
const match = await em.findOneOrFail(Match, {}, ['site']);
console.log(match.site.name); // no unwrap needed
console.log(match.site.creator.name); // property name does not exist on type reference
```

**Comment by vimmerru** [comment]
@darkbasic See here some PoC https://github.com/mikro-orm/mikro-orm/issues/214

**Comment by B4nan** [comment]
I believe we can close this in favour of #214. Which I am [planning to resolve in v4](https://github.com/mikro-orm/mikro-orm/issues/214#issuecomment-635480833).

**Comment by lookfirst** [comment]
Cool, no problem!
